File: src/features/group-quiz/services/group-quiz-service.ts
import { supabase } from "@/integrations/supabase/client";
import { generateQuestionsWithRetry } from "@/features/recommendations/services/ai-service";
import { generateRecommendations } from "@/features/recommendations/services/ai-service";
import { databaseService } from "@/features/recommendations/services/database-service";
import type { Answer } from "@/features/quiz/types/answer";
import type { Question } from "@/features/quiz/types/question";

import type {
  CreateSessionInput,
  GroupQuizResult,
  JoinSessionInput,
  QuizAnswer,
  QuizParticipant,
  QuizSession,
  QuizSessionStatus,
} from "../types/group-quiz";

const GUEST_ID_KEY = "smart-advisor.group-quiz.guest-participant-id";

const sessionsTable = () =>
  (supabase as unknown as { from: typeof supabase.from }).from(
    "quiz_sessions",
  );
const participantsTable = () =>
  (supabase as unknown as { from: typeof supabase.from }).from(
    "quiz_participants",
  );
const answersTable = () =>
  (supabase as unknown as { from: typeof supabase.from }).from("quiz_answers");

/** A 6-char shareable code without easily-confused characters. */
const generateCode = () => {
  const alphabet = "ABCDEFGHJKMNPQRSTUVWXYZ23456789";
  let code = "";
  for (let i = 0; i < 6; i += 1) {
    code += alphabet[Math.floor(Math.random() * alphabet.length)];
  }
  return code;
};

const guestParticipantId = (sessionCode: string) => {
  if (typeof window === "undefined") return null;
  const raw = window.localStorage.getItem(`${GUEST_ID_KEY}:${sessionCode}`);
  return raw ?? null;
};

const setGuestParticipantId = (sessionCode: string, id: string) => {
  if (typeof window === "undefined") return;
  window.localStorage.setItem(`${GUEST_ID_KEY}:${sessionCode}`, id);
};

class GroupQuizService {
  /**
   * Create a session as the host. The host must be authenticated (RLS).
   * Returns both the session and the host's participant row so the UI can
   * immediately treat the creator as a player too.
   */
  async createSession(input: CreateSessionInput): Promise<{
    session: QuizSession | null;
    participant: QuizParticipant | null;
    error: string | null;
  }> {
    const { data: authData } = await supabase.auth.getUser();
    if (!authData.user) {
      return {
        session: null,
        participant: null,
        error: "You must be signed in to host a group quiz.",
      };
    }

    // Try a few times in the (rare) case of a code collision.
    let session: QuizSession | null = null;
    for (let attempt = 0; attempt < 5; attempt += 1) {
      const code = generateCode();
      const { data, error } = (await sessionsTable()
        .insert({
          code,
          host_user_id: authData.user.id,
          status: "lobby",
          content_type: input.content_type,
          question_count: input.question_count,
        })
        .select("*")
        .single()) as { data: QuizSession | null; error: { code?: string; message: string } | null };

      if (!error && data) {
        session = data;
        break;
      }
      if (error?.code !== "23505") {
        return { session: null, participant: null, error: error?.message ?? "Failed to create session" };
      }
    }

    if (!session) {
      return {
        session: null,
        participant: null,
        error: "Couldn't generate a unique code, try again.",
      };
    }

    const { data: participant, error: pErr } = (await participantsTable()
      .insert({
        session_id: session.id,
        user_id: authData.user.id,
        display_name: input.display_name,
        is_host: true,
      })
      .select("*")
      .single()) as { data: QuizParticipant | null; error: { message: string } | null };

    if (pErr || !participant) {
      return {
        session,
        participant: null,
        error: pErr?.message ?? "Failed to register host as participant",
      };
    }
    return { session, participant, error: null };
  }

  /**
   * Look up a session by its share code, ignoring sessions that have already
   * passed their `expires_at` (24h after creation). Stale codes look like
   * "Session not found" rather than letting people resurrect expired lobbies.
   */
  async findByCode(
    code: string,
  ): Promise<{ session: QuizSession | null; error: string | null }> {
    const { data, error } = (await sessionsTable()
      .select("*")
      .eq("code", code.trim().toUpperCase())
      .gte("expires_at", new Date().toISOString())
      .maybeSingle()) as { data: QuizSession | null; error: { message: string } | null };
    if (error) return { session: null, error: error.message };
    return { session: data, error: null };
  }

  /**
   * Join a session — works for signed-in users and anonymous guests.
   * The guest's participant id is cached in localStorage so a refresh
   * doesn't kick them out.
   */
  async joinSession(input: JoinSessionInput): Promise<{
    session: QuizSession | null;
    participant: QuizParticipant | null;
    error: string | null;
  }> {
    const { session, error: lookupErr } = await this.findByCode(input.code);
    if (lookupErr || !session) {
      return { session: null, participant: null, error: lookupErr ?? "Code not found" };
    }
    if (session.status === "cancelled") {
      return { session: null, participant: null, error: "Session was cancelled." };
    }
    if (session.status === "completed") {
      return { session, participant: null, error: "Session already finished." };
    }

    // Re-use existing guest participant if present
    const existingGuestId = guestParticipantId(session.code);
    if (existingGuestId) {
      const { data: existing } = (await participantsTable()
        .select("*")
        .eq("id", existingGuestId)
        .maybeSingle()) as { data: QuizParticipant | null };
      if (existing && existing.session_id === session.id) {
        return { session, participant: existing, error: null };
      }
    }

    const { data: authData } = await supabase.auth.getUser();
    const userId = authData.user?.id ?? null;

    const { data: participant, error } = (await participantsTable()
      .insert({
        session_id: session.id,
        user_id: userId,
        display_name: input.display_name.trim() || "Guest",
        is_host: false,
      })
      .select("*")
      .single()) as { data: QuizParticipant | null; error: { message: string } | null };

    if (error || !participant) {
      return {
        session,
        participant: null,
        error: error?.message ?? "Failed to join session",
      };
    }

    if (!userId) setGuestParticipantId(session.code, participant.id);
    return { session, participant, error: null };
  }

  async listParticipants(
    sessionId: string,
  ): Promise<{ data: QuizParticipant[]; error: string | null }> {
    const { data, error } = (await participantsTable()
      .select("*")
      .eq("session_id", sessionId)
      .order("joined_at", { ascending: true })) as {
      data: QuizParticipant[] | null;
      error: { message: string } | null;
    };
    if (error) return { data: [], error: error.message };
    return { data: data ?? [], error: null };
  }

  async listAnswers(
    sessionId: string,
  ): Promise<{ data: QuizAnswer[]; error: string | null }> {
    const { data, error } = (await answersTable()
      .select("*")
      .eq("session_id", sessionId)) as {
      data: QuizAnswer[] | null;
      error: { message: string } | null;
    };
    if (error) return { data: [], error: error.message };
    return { data: data ?? [], error: null };
  }

  async submitAnswer(input: {
    session_id: string;
    participant_id: string;
    question_index: number;
    question: string;
    answer: string;
  }): Promise<{ error: string | null }> {
    const { error } = await answersTable()
      .upsert(input, { onConflict: "participant_id,question_index" })
      .select();
    return { error: error?.message ?? null };
  }

  async markParticipantSubmitted(
    participantId: string,
  ): Promise<{ error: string | null }> {
    const { error } = await participantsTable()
      .update({ answers_submitted_at: new Date().toISOString() })
      .eq("id", participantId);
    return { error: error?.message ?? null };
  }

  async setStatus(
    sessionId: string,
    status: QuizSessionStatus,
  ): Promise<{ error: string | null }> {
    const patch: Record<string, unknown> = { status };
    if (status === "completed") patch.completed_at = new Date().toISOString();
    const { error } = await sessionsTable().update(patch).eq("id", sessionId);
    return { error: error?.message ?? null };
  }

  async leave(participantId: string): Promise<{ error: string | null }> {
    const { error } = await participantsTable().delete().eq("id", participantId);
    return { error: error?.message ?? null };
  }

  /**
   * Pause the quiz and send everyone back to the lobby (e.g., the host wants
   * to wait for late joiners). Clears any in-flight answers, prior questions,
   * and the result so a fresh round starts cleanly when Start Quiz is
   * pressed again.
   */
  async returnToLobby(
    session: QuizSession,
  ): Promise<{ error: string | null }> {
    const { error: ansErr } = await answersTable()
      .delete()
      .eq("session_id", session.id);
    if (ansErr) return { error: ansErr.message };

    const { error: pErr } = await participantsTable()
      .update({ answers_submitted_at: null })
      .eq("session_id", session.id);
    if (pErr) return { error: pErr.message };

    const { error: sErr } = await sessionsTable()
      .update({
        status: "lobby",
        questions: null,
        result: null,
        completed_at: null,
      })
      .eq("id", session.id);
    return { error: sErr?.message ?? null };
  }

  /**
   * Reset the session for another round with the same lobby — clear all
   * answers, reset every participant's submitted timestamp, drop the prior
   * result, then generate fresh questions and flip status back to in_progress.
   */
  async restartSession(
    session: QuizSession,
    hostAge: number,
    hostName: string,
  ): Promise<{ questions: Question[] | null; error: string | null }> {
    const { error: ansErr } = await answersTable()
      .delete()
      .eq("session_id", session.id);
    if (ansErr) return { questions: null, error: ansErr.message };

    const { error: pErr } = await participantsTable()
      .update({ answers_submitted_at: null })
      .eq("session_id", session.id);
    if (pErr) return { questions: null, error: pErr.message };

    const { error: sErr } = await sessionsTable()
      .update({ result: null, completed_at: null })
      .eq("id", session.id);
    if (sErr) return { questions: null, error: sErr.message };

    return this.generateAndStartQuiz(session, hostAge, hostName);
  }

  /**
   * Generate the question set for a session and persist it on the row, then
   * flip status to in_progress. All participants pick up the questions via
   * the realtime subscription on `quiz_sessions`.
   */
  async generateAndStartQuiz(
    session: QuizSession,
    hostAge: number,
    hostName: string,
  ): Promise<{ questions: Question[] | null; error: string | null }> {
    try {
      const questions = await generateQuestionsWithRetry(
        session.content_type,
        hostAge,
        session.question_count,
        hostName,
      );
      const { error } = await sessionsTable()
        .update({ questions, status: "in_progress" })
        .eq("id", session.id);
      if (error) return { questions: null, error: error.message };
      return { questions, error: null };
    } catch (err) {
      return {
        questions: null,
        error:
          err instanceof Error
            ? err.message
            : "Failed to generate questions",
      };
    }
  }

  /**
   * Combine every participant's answers into a single synthetic answer set
   * and ask the existing recommendations edge function to find common ground.
   * Saves the result on the session and flips status to completed.
   */
  async synthesizeRecommendation(
    session: QuizSession,
    participants: QuizParticipant[],
    answers: QuizAnswer[],
    hostAge: number,
    hostName: string,
  ): Promise<{ result: GroupQuizResult | null; error: string | null }> {
    if (!session.questions || session.questions.length === 0) {
      return { result: null, error: "No questions on session" };
    }

    // Group answers by question_index → list of `Name: answer` lines
    const byQuestion = new Map<number, string[]>();
    const nameById = new Map(
      participants.map((p) => [p.id, p.display_name] as const),
    );
    answers.forEach((a) => {
      const name = nameById.get(a.participant_id) ?? "Player";
      const list = byQuestion.get(a.question_index) ?? [];
      list.push(`${name}: ${a.answer}`);
      byQuestion.set(a.question_index, list);
    });

    const combinedAnswers: Answer[] = session.questions.map((q, idx) => ({
      id: `combined-${idx}`,
      question_id: q.id,
      question_text: q.text,
      answer_text:
        byQuestion.get(idx)?.join(" | ") ??
        "(no responses — pick something widely appealing)",
      created_at: new Date().toISOString(),
    }));

    try {
      const data = await generateRecommendations(
        combinedAnswers,
        session.content_type,
        hostAge,
        // Synthesizer prompt: tell the model these are multiple players.
        `${hostName} hosting a group of ${participants.length}`,
      );
      const result: GroupQuizResult = {};
      if (data.movieRecommendation) {
        const m = data.movieRecommendation;
        result.movie = {
          title: m.title,
          director: m.director,
          year: m.year,
          genres: m.genres,
          explanation: m.explanation,
        };
      }
      if (data.bookRecommendation) {
        const b = data.bookRecommendation;
        result.book = {
          title: b.title,
          author: b.author,
          year: b.year,
          genres: b.genres,
          explanation: b.explanation,
        };
      }

      // Persist the picks to the host's recommendations table so they show
      // up in History/Library and feed future taste signals. We swallow any
      // save errors — the synth result is already committed to the session
      // row, so the group experience isn't blocked by a save failure.
      // saveRecommendation reads user_id from the auth session, so user_id
      // here is a stub the service ignores — but the type requires it.
      let savedRecommendationId: string | null = null;
      if (result.movie) {
        const saved = await databaseService.saveRecommendation({
          user_id: "",
          type: "movie",
          title: result.movie.title,
          director: result.movie.director,
          year: result.movie.year,
          genres: result.movie.genres ?? [],
          explanation: result.movie.explanation,
          content_type: session.content_type,
          is_favorited: false,
          rating: 0,
        } as Parameters<typeof databaseService.saveRecommendation>[0]);
        savedRecommendationId = saved.data?.id ?? null;
      }
      if (result.book) {
        const saved = await databaseService.saveRecommendation({
          user_id: "",
          type: "book",
          title: result.book.title,
          author: result.book.author,
          year: result.book.year,
          genres: result.book.genres ?? [],
          explanation: result.book.explanation,
          content_type: session.content_type,
          is_favorited: false,
          rating: 0,
        } as Parameters<typeof databaseService.saveRecommendation>[0]);
        savedRecommendationId = savedRecommendationId ?? saved.data?.id ?? null;
      }

      const { error } = await sessionsTable()
        .update({
          result,
          status: "completed",
          completed_at: new Date().toISOString(),
          recommendation_id: savedRecommendationId,
        })
        .eq("id", session.id);
      if (error) return { result, error: error.message };
      return { result, error: null };
    } catch (err) {
      return {
        result: null,
        error:
          err instanceof Error
            ? err.message
            : "Failed to synthesize recommendation",
      };
    }
  }
}

export const groupQuizService = new GroupQuizService();
export { guestParticipantId, setGuestParticipantId };

File: apps/mobile/lib/features/group_quiz/services/group_quiz_service.dart
import 'dart:math';

import 'package:shared_preferences/shared_preferences.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../../core/models/ai_models.dart';
import '../../../core/models/answer.dart';
import '../../../core/models/enums.dart';
import '../../../core/models/question.dart';
import '../../../core/result.dart';
import '../../recommendations/services/ai_service.dart';
import '../models/group_quiz.dart';

/// Port of web group-quiz-service.ts. All direct Supabase (no Next.js
/// routes); questions/recs via the anthropic-* Edge Functions through
/// AiService. Realtime is handled separately by the session provider.
class GroupQuizService {
  GroupQuizService(this._c, this._ai);

  final SupabaseClient _c;
  final AiService _ai;

  static const _sessions = 'quiz_sessions';
  static const _participants = 'quiz_participants';
  static const _answers = 'quiz_answers';

  String _code() {
    const a = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
    final r = Random.secure();
    return List.generate(6, (_) => a[r.nextInt(a.length)]).join();
  }

  Future<({QuizSession? session, QuizParticipant? participant, String? error})>
      createSession({
    required String contentType,
    required int questionCount,
    required int maxParticipants,
    required String displayName,
    // Async mode (additive): live callers omit both and behaviour is
    // identical to before. A non-null deadlineAt creates an async session.
    DateTime? deadlineAt,
    DateTime? plannedFor,
  }) async {
    final uid = _c.auth.currentUser?.id;
    if (uid == null) {
      return (session: null, participant: null, error: 'Sign in to host.');
    }
    try {
      final s = await _c
          .from(_sessions)
          .insert({
            'code': _code(),
            'host_user_id': uid,
            'content_type': contentType,
            'question_count': questionCount,
            'max_participants': maxParticipants,
            if (deadlineAt != null)
              'deadline_at': deadlineAt.toUtc().toIso8601String(),
            if (plannedFor != null)
              'planned_for': plannedFor.toUtc().toIso8601String(),
          })
          .select()
          .single();
      final session = QuizSession.fromJson(Map<String, dynamic>.from(s));
      final p = await _c
          .from(_participants)
          .insert({
            'session_id': session.id,
            'user_id': uid,
            'display_name': displayName,
            'is_host': true,
          })
          .select()
          .single();
      return (
        session: session,
        participant:
            QuizParticipant.fromJson(Map<String, dynamic>.from(p)),
        error: null,
      );
    } on PostgrestException catch (e) {
      return (session: null, participant: null, error: e.message);
    }
  }

  Future<({QuizSession? session, String? error})> findByCode(
      String code) async {
    try {
      final row = await _c
          .from(_sessions)
          .select()
          .eq('code', code.toUpperCase())
          .gte('expires_at', DateTime.now().toUtc().toIso8601String())
          .maybeSingle();
      if (row == null) {
        return (session: null, error: 'Session not found or expired.');
      }
      return (
        session: QuizSession.fromJson(Map<String, dynamic>.from(row)),
        error: null
      );
    } on PostgrestException catch (e) {
      return (session: null, error: e.message);
    }
  }

  Future<({QuizSession? session, QuizParticipant? participant, String? error})>
      joinSession(
          {required String code, required String displayName}) async {
    final found = await findByCode(code);
    if (found.session == null) {
      return (session: null, participant: null, error: found.error);
    }
    final session = found.session!;
    try {
      final count = await _c
          .from(_participants)
          .select('id')
          .eq('session_id', session.id)
          .count();
      if (count.count >= session.maxParticipants) {
        return (session: null, participant: null, error: 'Session is full.');
      }
      final p = await _c
          .from(_participants)
          .insert({
            'session_id': session.id,
            'user_id': _c.auth.currentUser?.id,
            'display_name': displayName,
          })
          .select()
          .single();
      final participant =
          QuizParticipant.fromJson(Map<String, dynamic>.from(p));
      await _setGuestId(session.code, participant.id);
      return (session: session, participant: participant, error: null);
    } on PostgrestException catch (e) {
      return (session: null, participant: null, error: e.message);
    }
  }

  Future<QuizSession?> getSession(String sessionId) async {
    final row = await _c
        .from(_sessions)
        .select()
        .eq('id', sessionId)
        .maybeSingle();
    return row == null
        ? null
        : QuizSession.fromJson(Map<String, dynamic>.from(row));
  }

  Future<List<QuizParticipant>> listParticipants(String sessionId) async {
    final rows = await _c
        .from(_participants)
        .select()
        .eq('session_id', sessionId)
        .order('joined_at');
    return rows
        .map((r) => QuizParticipant.fromJson(Map<String, dynamic>.from(r)))
        .toList();
  }

  Future<List<QuizAnswerRow>> listAnswers(String sessionId) async {
    final rows =
        await _c.from(_answers).select().eq('session_id', sessionId);
    return rows
        .map((r) => QuizAnswerRow.fromJson(Map<String, dynamic>.from(r)))
        .toList();
  }

  Future<ServiceResult<void>> submitAnswer({
    required String sessionId,
    required String participantId,
    required int questionIndex,
    required String question,
    required String answer,
  }) async {
    try {
      await _c.from(_answers).upsert({
        'session_id': sessionId,
        'participant_id': participantId,
        'question_index': questionIndex,
        'question': question,
        'answer': answer,
      }, onConflict: 'participant_id,question_index');
      return ServiceResult.ok(null);
    } on PostgrestException catch (e) {
      return ServiceResult.fail(e.message);
    }
  }

  Future<ServiceResult<void>> markParticipantSubmitted(
      String participantId) async {
    try {
      await _c.from(_participants).update({
        'answers_submitted_at': DateTime.now().toUtc().toIso8601String(),
      }).eq('id', participantId);
      return ServiceResult.ok(null);
    } on PostgrestException catch (e) {
      return ServiceResult.fail(e.message);
    }
  }

  Future<ServiceResult<void>> setStatus(
      String sessionId, QuizSessionStatus status) async {
    try {
      await _c.from(_sessions).update({
        'status': status.wire,
        if (status == QuizSessionStatus.completed)
          'completed_at': DateTime.now().toUtc().toIso8601String(),
      }).eq('id', sessionId);
      return ServiceResult.ok(null);
    } on PostgrestException catch (e) {
      return ServiceResult.fail(e.message);
    }
  }

  Future<ServiceResult<void>> leave(String participantId) async {
    try {
      await _c.from(_participants).delete().eq('id', participantId);
      return ServiceResult.ok(null);
    } on PostgrestException catch (e) {
      return ServiceResult.fail(e.message);
    }
  }

  Future<({List<Question>? questions, String? error})> generateAndStartQuiz(
      QuizSession session, int hostAge, String hostName,
      {required String contentTone}) async {
    try {
      final qs = await _ai.generateQuestionsWithRetry(
        contentType: ContentType.fromWire(session.contentType),
        userAge: hostAge,
        contentTone: contentTone,
        questionCount: session.questionCount,
        userName: hostName,
      );
      await _c.from(_sessions).update({
        'questions': qs.map((q) => q.toJson()).toList(),
        'status': QuizSessionStatus.inProgress.wire,
      }).eq('id', session.id);
      return (questions: qs, error: null);
    } on AiServiceException catch (e) {
      return (questions: null, error: e.message);
    }
  }

  /// Combine every participant's answers into one profile and synthesize a
  /// shared recommendation, then complete the session (realtime propagates).
  ///
  /// This is the live-mode completion entry point and is unchanged in
  /// behaviour — it just delegates the actual synthesis+finalize to the
  /// shared [_finalize] helper so async mode can reuse the exact same
  /// recommendation/completion logic instead of duplicating it.
  Future<ServiceResult<void>> synthesizeRecommendation(
      QuizSession session, List<QuizParticipant> participants, int hostAge,
      String hostName,
      {required String contentTone}) {
    return _finalize(
      session,
      participants.length,
      hostAge,
      hostName,
      contentTone: contentTone,
    );
  }

  /// Shared synthesize-and-complete used by BOTH live completion and async
  /// resolution. Sets result/status/completed_at exactly as live mode did.
  Future<ServiceResult<void>> _finalize(
      QuizSession session, int groupSize, int hostAge, String hostName,
      {required String contentTone}) async {
    final answerRows = await listAnswers(session.id);
    final answers = [
      for (var i = 0; i < answerRows.length; i++)
        Answer(
          id: 'g$i',
          questionId: 'q${answerRows[i].questionIndex}',
          questionText: answerRows[i].question,
          answerText: answerRows[i].answer,
        ),
    ];
    try {
      final data = await _ai.generateRecommendationsWithRetry(
        answers: answers,
        contentType: ContentType.fromWire(session.contentType),
        userAge: hostAge,
        contentTone: contentTone,
        userName: '$hostName hosting $groupSize group',
      );
      GroupQuizPick? pick(item) => item == null
          ? null
          : GroupQuizPick(
              title: item.title,
              director: item.director,
              author: item.author,
              artist: item.artist,
              year: item.year,
              genres: item.genres,
              explanation: item.explanation,
              description: item.description,
            );
      final result = GroupQuizResult(
        movie: pick(data.movieRecommendation),
        book: pick(data.bookRecommendation),
        music: pick(data.musicRecommendation),
      );
      await _c.from(_sessions).update({
        'result': result.toJson(),
        'status': QuizSessionStatus.completed.wire,
        'completed_at': DateTime.now().toUtc().toIso8601String(),
      }).eq('id', session.id);
      return ServiceResult.ok(null);
    } on AiServiceException catch (e) {
      return ServiceResult.fail(e.message);
    }
  }

  /// Async-mode resolution without a server scheduler. Resolves the group
  /// pick CLIENT-side when either condition holds:
  ///   * every joined participant has submitted, OR
  ///   * the deadline has passed (now >= deadline_at).
  /// Idempotent — guarded on a non-completed async session, so concurrent
  /// callers (last submitter + a post-deadline opener) can't double-run and
  /// the live (deadline_at == null) path is never touched.
  ///
  /// Returns (resolved, error): resolved == true only when this call
  /// actually performed the finalize; false means "not ready / already
  /// done" (not an error).
  Future<({bool resolved, String? error})> resolveIfReady(
      String sessionId, int hostAge, String hostName,
      {required String contentTone}) async {
    final session = await getSession(sessionId);
    if (session == null) {
      return (resolved: false, error: 'Session not found.');
    }
    // Live sessions (no deadline) and already-completed sessions are
    // explicitly out of scope — never alter the live completion path.
    if (!session.isAsync ||
        session.status == QuizSessionStatus.completed) {
      return (resolved: false, error: null);
    }
    final parts = await listParticipants(sessionId);
    final allIn = parts.isNotEmpty &&
        parts.every((p) => p.answersSubmittedAt != null);
    final deadline = session.deadlineAtUtc;
    final past = deadline != null &&
        !DateTime.now().toUtc().isBefore(deadline);
    if (!allIn && !past) {
      return (resolved: false, error: null);
    }
    final r = await _finalize(
      session,
      parts.length,
      hostAge,
      hostName,
      contentTone: contentTone,
    );
    return (resolved: !r.isError, error: r.error);
  }

  /// Async create flow: generate the question set and move the session
  /// straight to in_progress (no host-driven lobby start), reusing the
  /// same question generation as live. Live mode keeps using
  /// generateAndStartQuiz from the lobby — that path is unchanged.
  Future<({List<Question>? questions, String? error})> startAsyncQuiz(
      QuizSession session, int hostAge, String hostName,
      {required String contentTone}) {
    return generateAndStartQuiz(session, hostAge, hostName,
        contentTone: contentTone);
  }

  // Guest participant id cache (parity with web localStorage).
  Future<void> _setGuestId(String code, String id) async {
    final p = await SharedPreferences.getInstance();
    await p.setString('sa.group-quiz.guest:$code', id);
  }

  Future<String?> guestParticipantId(String code) async {
    final p = await SharedPreferences.getInstance();
    return p.getString('sa.group-quiz.guest:$code');
  }
}
